dataframely/collection/collection.py

```diff
@@ -370,29 +372,86 @@ def validate(cls, data: Mapping[str, FrameType], /, *, cast: bool = False) -> Se
+                "shallow" lazy frames, i.e. lazy frames by simply calling
+                :meth:`~polars.DataFrame.lazy` on the validated data frame. If
+                ``False``, this method only raises if ``data`` does not contain data for
+                all required members. The returned collection contains "true" lazy
+                frames that will be validated upon calling
+                :meth:`~polars.LazyFrame.collect` on the individual member or
+                :meth:`collect_all` on the collection. Note that, in the latter case,
+                information from error messages is limited.
 
         Raises:
             ValueError: If an insufficient set of input data frames is provided, i.e. if
                 any required member of this collection is missing in the input.
-            ValidationError: If any of the input data frames does not satisfy its schema
-                definition or the filters on this collection result in the removal of at
-                least one row across any of the input data frames.
+            ComputeError: If ``eager=True`` and any of the input data frames does not
+                satisfy its schema definition or the filters on this collection result
+                in the removal of at least one row across any of the input data frames.
 
         Returns:
             An instance of the collection. All members of the collection are guaranteed
             to be valid with respect to their respective schemas and the filters on this
             collection did not remove rows from any member. The input order of each
             member is maintained.
         """
-        out, failure = cls.filter(data, cast=cast)
-        if any(len(fail) > 0 for fail in failure.values()):
-            raise MemberValidationError(
-                {
-                    name: RuleValidationError(fail.counts())
-                    for name, fail in failure.items()
+        cls._validate_input_keys(data)
+
+        if eager:
+            # If we perform the validation eagerly, we call filter and check the failure
+            # information to properly construct a useful error message.
+            filtered, failures = cls.filter(data, cast=cast, eager=True)
+            if any(len(failure) > 0 for failure in failures.values()):
+                errors = {
+                    member: format_rule_failures(list(failure.counts().items()))
+                    for member, failure in failures.items()
+                    if len(failure) > 0
                 }
-            )
-        return out
+                details = [
+                    f" > Member '{member}' failed validation:\n"
+                    + textwrap.indent(error, "   ")
+                    for member, error in errors.items()
+                ]
+                message = "\n".join(
+                    [f"{len(errors)} members failed validation:"] + details
+                )
+                raise plexc.ComputeError(message)
```

Comment:
I assume you chose `ComputeError` here in order to be consistent with the lazy case. Is that right? While consistency is good, I feel sad that I don't get a nice `ValidationError` type here. How bad would it be to break consistency and raise a `ValidationError` in this branch? Consistency would mainly be useful for users who can use a single code path for both lazy and eager validation, but is that even a possible outcome? After all, even if it is the same exception type, I would have to deal with it in two different places (either when I call `validate` or when I call `collect`), which could be far apart. The more I think about it, the more I think that consistency is not really a concern here